### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [1]:
# Dependencies and Setup
import pandas as pd

# File to Load (Remember to Change These)
file_to_load = "Resources/purchase_data.csv"

# Read Purchasing File and store into Pandas data frame
purchase_data = pd.read_csv(file_to_load)

## Player Count

* Display the total number of players


In [2]:
total_players = purchase_data["SN"].nunique()

total_players_df = pd.DataFrame({
    "Total Players": total_players}, index=[0])

total_players_df

Total Players
0            576

## Purchasing Analysis (Total)

* Run basic calculations to obtain number of unique items, average price, etc.


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame


In [3]:
#discover the variables values

unique_items = purchase_data["Item Name"].nunique()
av_price = purchase_data["Price"].mean()
num_purchases = purchase_data["Purchase ID"].count()
total_revenue = purchase_data["Price"].sum()

#create a data fram with the new values

raw_info = {
    "unique_items": [unique_items],
    "av_price": [av_price],
    "num_purchases":[num_purchases],
    "total_revenue": [total_revenue]
}

summary_df = pd.DataFrame(raw_info, columns=[
    "unique_items",
    "av_price",
    "num_purchases",
    "total_revenue"])

#rename the columns

summary_df = summary_df.rename(columns={
    "unique_items":"Number of Unique Items",
    "av_price":"Average Price",
    "num_purchases":"Number of Purchases",
    "total_revenue": "Total Revenue"
})

#format the values

summary_df["Average Price"] = summary_df["Average Price"].map("${:,.2f}".format)
summary_df["Total Revenue"] = summary_df["Total Revenue"].map("${:,.2f}".format)

#print the results

summary_df

Number of Unique Items Average Price  Number of Purchases Total Revenue
0                     179         $3.05                  780     $2,379.77

## Gender Demographics

* Percentage and Count of Male Players


* Percentage and Count of Female Players


* Percentage and Count of Other / Non-Disclosed




In [4]:
#group by Gender

gender_group = purchase_data.groupby("Gender")

purchase_data["Gender"]

#Using the group, count the number of UNIQUE Screennames

total_count = gender_group["SN"].nunique()

#Find the Percentage of players using the Unique number of screennames

percentage_players = ((total_count / total_count.sum()) *100)

#format the variables
percentage_players = percentage_players.map("{:.2f}%".format)

#Create a DataFrame using the Totals and Percentages

gender_demo_df = pd.DataFrame({"Total Count": total_count,
                               "Percentage of Players":percentage_players
                              })

gender_demo_df = gender_demo_df.reset_index()

gender_demo_df = gender_demo_df.set_index("Gender")

gender_demo_df.sort_values("Total Count", ascending=False)

Total Count Percentage of Players
Gender                                                  
Male                           484                84.03%
Female                          81                14.06%
Other / Non-Disclosed           11                 1.91%


## Purchasing Analysis (Gender)

* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender




* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [5]:
#group by Gender

gender_group = purchase_data.groupby("Gender")

#Using this group, count the number of times a purchase was made

total_count_purchase = gender_group["Purchase ID"].count()

#Using the group, find the total sum of the amount spent

unique_person_totals = gender_group["Price"].sum()

#Using the group, find the total count of the UNIQUE people making perchases

unique_person_count = gender_group["SN"].nunique()

#Using the group, find the average of the amount spent

av_purchase = gender_group["Price"].mean()

#Using the group, find the total of the amount spent

Total_Value = gender_group["Price"].sum()

#Using the group, find the average of the amount spent per person

Per_Person_Total = unique_person_totals / unique_person_count

#format your variables and results

Total_Value = Total_Value.map("${:.2f}".format)
av_purchase = av_purchase.map("${:.2f}".format)
Per_Person_Total = Per_Person_Total.map("${:.2f}".format)

#Create a DataFrame using created dictionaries

summary_purchase_df = pd.DataFrame({"Purchase Count" : total_count_purchase,
                                   "Average Purchase Price" : av_purchase,
                                   "Total Purchase Value": Total_Value,
                                   "Avg Total Purchase per Person" : Per_Person_Total})

summary_purchase_df = summary_purchase_df.reset_index()
summary_purchase_df = summary_purchase_df.set_index("Gender")
summary_purchase_df.head()

Purchase Count Average Purchase Price  \
Gender                                                         
Female                            113                  $3.20   
Male                              652                  $3.02   
Other / Non-Disclosed              15                  $3.35   

                      Total Purchase Value Avg Total Purchase per Person  
Gender                                                                    
Female                             $361.94                         $4.47  
Male                              $1967.64                         $4.07  
Other / Non-Disclosed               $50.19                         $4.56

## Age Demographics

* Establish bins for ages


* Categorize the existing players using the age bins. Hint: use pd.cut()


* Calculate the numbers and percentages by age group


* Create a summary data frame to hold the results


* Optional: round the percentage column to two decimal points


* Display Age Demographics Table


In [6]:
bins = [0, 9, 14, 19, 24, 29, 34, 39, 100]

bin_names = ["<10", "10-14", "15-19", "20-24", "25-29", "30-34", "35-39", "40+"]

purchase_data["Age Brackets"] = pd.cut(purchase_data["Age"], bins, labels=bin_names, include_lowest=True)

In [7]:
total_count_age_group = purchase_data.groupby("Age Brackets").nunique()

per_bracket_total = total_count_age_group["SN"]

total_purchases = purchase_data["SN"].nunique()

percentage_players = (per_bracket_total / total_purchases) *100

percentage_players = percentage_players.map("{:.2f}%".format)

age_bracket_df = pd.DataFrame({"Total Count" : per_bracket_total,
                              "Percentage of Players": percentage_players})

age_bracket_df = age_bracket_df.reset_index()
age_bracket_df = age_bracket_df.set_index("Age Brackets")
age_bracket_df                               

Total Count Percentage of Players
Age Brackets                                   
<10                    17                 2.95%
10-14                  22                 3.82%
15-19                 107                18.58%
20-24                 258                44.79%
25-29                  77                13.37%
30-34                  52                 9.03%
35-39                  31                 5.38%
40+                    12                 2.08%

## Purchasing Analysis (Age)

* Bin the purchase_data data frame by age


* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [8]:
#purchase_data.head()

age_group = purchase_data.groupby("Age Brackets")

purchase_count = age_group["SN"].count()

age_person_totals = age_group["Price"].sum()
age_person_count = age_group["SN"].nunique()

age_av_purchase = age_group["Price"].mean()

age_total_value = age_group["Price"].sum()

age_per_person_total = age_person_totals / age_person_count

age_total_value = age_total_value.map("${:.2f}".format)
age_av_purchase = age_av_purchase.map("${:.2f}".format)
age_per_person_total = age_per_person_total.map("${:.2f}".format)

age_summary_purchase_df = pd.DataFrame({"Purchase Count" : purchase_count,
                                   "Average Purchase Price" : age_av_purchase,
                                   "Total Purchase Value": age_total_value,
                                   "Avg Total Purchase per Person" : age_per_person_total})

age_summary_purchase_df = age_summary_purchase_df.reset_index()

age_summary_purchase_df = age_summary_purchase_df.rename(columns={"Age Brackets": "Age Ranges"})

age_summary_purchase_df = age_summary_purchase_df.set_index("Age Ranges")

age_summary_purchase_df

Purchase Count Average Purchase Price Total Purchase Value  \
Age Ranges                                                               
<10                     23                  $3.35               $77.13   
10-14                   28                  $2.96               $82.78   
15-19                  136                  $3.04              $412.89   
20-24                  365                  $3.05             $1114.06   
25-29                  101                  $2.90              $293.00   
30-34                   73                  $2.93              $214.00   
35-39                   41                  $3.60              $147.67   
40+                     13                  $2.94               $38.24   

           Avg Total Purchase per Person  
Age Ranges                                
<10                                $4.54  
10-14                              $3.76  
15-19                              $3.86  
20-24                              $4.32  
25-29                              $3.81  
30-34                              $4.12  
35-39                              $4.76  
40+                                $3.19

## Top Spenders

* Run basic calculations to obtain the results in the table below


* Create a summary data frame to hold the results


* Sort the total purchase value column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [9]:
screennames = purchase_data.groupby("SN")

sn_purchase_count = screennames["SN"].count()

sn_person_totals = screennames["Price"].sum()
sn_person_count = screennames["SN"].nunique()

sn_av_purchase = screennames["Price"].mean()

sn_total_value = screennames["Price"].sum()

#Put into a table
sn_summary_purchase_df = pd.DataFrame({"Purchase Count" : sn_purchase_count,
                                   "Average Purchase Price" : sn_av_purchase,
                                   "Total Purchase Value": sn_total_value
                                   })

sn_summary_purchase_df = sn_summary_purchase_df.sort_values(["Total Purchase Value"],ascending=False)

#Format the Data
sn_summary_purchase_df["Total Purchase Value"] = sn_summary_purchase_df["Total Purchase Value"].map("${:.2f}".format)
sn_summary_purchase_df["Average Purchase Price"] = sn_summary_purchase_df["Average Purchase Price"].map("${:.2f}".format)

sn_summary_purchase_df.head(5)

Purchase Count Average Purchase Price Total Purchase Value
SN                                                                     
Lisosia93                 5                  $3.79               $18.96
Idastidru52               4                  $3.86               $15.45
Chamjask73                3                  $4.61               $13.83
Iral74                    4                  $3.40               $13.62
Iskadarya95               3                  $4.37               $13.10

## Most Popular Items

* Retrieve the Item ID, Item Name, and Item Price columns


* Group by Item ID and Item Name. Perform calculations to obtain purchase count, average item price, and total purchase value


* Create a summary data frame to hold the results


* Sort the purchase count column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [10]:
item_id = purchase_data.groupby(["Item ID","Item Name"])

id_purchase_count = item_id["Purchase ID"].count()

id_item_price = item_id["Price"].mean()

id_total_revenue = item_id["Price"].sum()

item_id_summary_df = pd.DataFrame({"Purchase Count" : id_purchase_count,
                                  "Item Price": id_item_price,
                                  "Total Purchase Value" : id_total_revenue})

item_id_summary_df = item_id_summary_df.sort_values(["Purchase Count"],ascending=False)

#Format the Data
item_id_summary_df["Item Price"] = item_id_summary_df["Item Price"].map("${:.2f}".format)
item_id_summary_df["Total Purchase Value"] = item_id_summary_df["Total Purchase Value"].map("${:.2f}".format)

item_id_summary_df.head()

Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
145     Fiery Glass Crusader                                       9   
132     Persuasion                                                 9   
108     Extraction, Quickblade Of Trembling Hands                  9   

                                                     Item Price  \
Item ID Item Name                                                 
92      Final Critic                                      $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
145     Fiery Glass Crusader                              $4.58   
132     Persuasion                                        $3.22   
108     Extraction, Quickblade Of Trembling Hands         $3.53   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
145     Fiery Glass Crusader                                       $41.22  
132     Persuasion                                                 $28.99  
108     Extraction, Quickblade Of Trembling Hands                  $31.77

## Most Profitable Items

* Sort the above table by total purchase value in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the data frame



In [11]:
item_id_summary_df = pd.DataFrame({"Purchase Count" : id_purchase_count,
                                   "Item Price": id_item_price,
                                   "Total Purchase Value" : id_total_revenue})

item_id_summary_df = item_id_summary_df.sort_values(["Total Purchase Value"],ascending=False)

#Format the Data
item_id_summary_df["Item Price"] = item_id_summary_df["Item Price"].map("${:.2f}".format)
item_id_summary_df["Total Purchase Value"] = item_id_summary_df["Total Purchase Value"].map("${:.2f}".format)

item_id_summary_df.head()

Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
82      Nirvana                                                    9   
145     Fiery Glass Crusader                                       9   
103     Singed Scalpel                                             8   

                                                     Item Price  \
Item ID Item Name                                                 
92      Final Critic                                      $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
82      Nirvana                                           $4.90   
145     Fiery Glass Crusader                              $4.58   
103     Singed Scalpel                                    $4.35   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
82      Nirvana                                                    $44.10  
145     Fiery Glass Crusader                                       $41.22  
103     Singed Scalpel                                             $34.80

## You must include a written description of at least two observable trends based on the data.


Observation 1:

The core demographic of this game that purchases in game items are 20-24, seconded by 15-19 year olds.

Observation 2:

The VAST majority of the players are male (Male 84.03%)

Observation 3:

Females tend to spend more money per purchase then males ($3.20)